cmd/tailscale/cli: add `remove` command for accounts

Esteban-Bermudez opened the pull request:
Fixes #12255

Add a new command to the CLI to remove a profile from the local client. This also fixes a bug in the local client where the DELETE request was not being sent correctly.

I decided to go with this approach instead of what is mentioned as a request solution for the following reasons.

1. Since commands like `switch`, `login` and `logout` reference the user's account(s), I felt it made more sense to have it be a regular command instead of the functionality being contained under the `switch` command. 
2. This makes it more accessible for Linux users who do not have a GUI interface and would look towards `tailscale --help` for directions on how to remove an account
3. The command name `remove` seemed a bit clearer than using `delete` as `delete` could be interpreted as fully purging an account and its info. Additionally, `remove` is the word chosen for the GUI options

zhangguanzhang commented:
@bradfitz PTAL, this will useful for multi server of headscale

Esteban-Bermudez commented:
Rebased with the latest v1.82.0 changes

bradfitz commented on cmd/tailscale/cli/remove.go:
```diff
@@ -0,0 +1,159 @@
+
+var removeCmd = &ffcli.Command{
+	Name:       "remove",
+	ShortUsage: "tailscale remove <id>",
```
Can we put this "remove" as a subcommand under something else? I'd rather it not be so prominent in the top-level docs.

bradfitz commented on cmd/tailscale/cli/cli.go:
```diff
@@ -191,6 +191,7 @@ change in the future.
 			loginCmd,
 			logoutCmd,
 			switchCmd,
+           removeCmd,
```
gofmt

bradfitz commented on client/local/local.go:
```diff
@@ -1504,7 +1504,7 @@ func (lc *Client) SwitchProfile(ctx context.Context, profile ipn.ProfileID) erro
 // will be selected as if SwitchToEmptyProfile was called.
 func (lc *Client) DeleteProfile(ctx context.Context, profile ipn.ProfileID) error {
-	_, err := lc.send(ctx, "DELETE", "/localapi/v0/profiles"+url.PathEscape(string(profile)), http.StatusNoContent, nil)
+	_, err := lc.send(ctx, "DELETE", "/localapi/v0/profiles/"+url.PathEscape(string(profile)), http.StatusNoContent, nil)
```
wow, whoops. This could even be its own PR if you wanted.

Esteban-Bermudez replied:
I will create a separate PR for this and @ you.

Esteban-Bermudez replied:
@bradfitz https://github.com/tailscale/tailscale/pull/15482

bradfitz commented on cmd/tailscale/cli/remove.go:
```diff
@@ -0,0 +1,159 @@
+			break
+		}
+	}
+	fmt.Println(profID)
```
are these leftover debug prints?

bradfitz commented on cmd/tailscale/cli/remove.go:
```diff
@@ -0,0 +1,159 @@
+		os.Exit(1)
+	}
+
+	printf("Removing account %q\n", args[0])
```
https://www.linfo.org/rule_of_silence.html ?

bradfitz commented on cmd/tailscale/cli/remove.go:
```diff
@@ -0,0 +1,159 @@
+	}
+
+	printf("Removing account %q\n", args[0])
+	for {
```
why is this loop needed? We can't just do the DeleteProfile call and return?

Esteban-Bermudez replied:
I am not sure, I based this part of the code on the functionality found in `switch.go`

https://github.com/tailscale/tailscale/blob/a7be3a3d86fe4bcde169db1c6e4f0e398c887498/cmd/tailscale/cli/switch.go#L146-L179

techster10173 commented:
What's the current state of this PR? This is very much needed and would like to help push this forward

Esteban-Bermudez commented on cmd/tailscale/cli/switch.go:
```diff
@@ -34,6 +34,27 @@ This command is currently in alpha and may change in the future.`,
+
+This command is currently in alpha and may change in the future.`,
+			FlagSet: func() *flag.FlagSet {
+				fs := flag.NewFlagSet("switch", flag.ExitOnError)
```
I have removed the flag under the subcommand

Esteban-Bermudez commented:
I have moved the functionality of the `remove` command to be a subcommand to `switch`. It could also make sense to have it under `logout`. I have also removed the for loop after removing the account so that it functions like the implementation for the `logout` command.

In the future it would be cool to see all the account related functionality under one root command called `account` which would hold commands like `login`, `logout` , `switch` and then eventually `remove`.

Esteban-Bermudez commented:
@bradfitz Are you able to look at this again so that it can be merged to fully resolve https://github.com/tailscale/tailscale/issues/12255? Thank you.

sailorfrag commented on cmd/tailscale/cli/switch.go:
```diff
@@ -178,3 +194,51 @@ func switchProfile(ctx context.Context, args []string) error {
+	var profID ipn.ProfileID
+	// Allow matching by ID, Tailnet, or Account
+	// in that order.
+	for _, p := range all {
+		if p.ID == ipn.ProfileID(args[0]) {
+			profID = p.ID
+			break
+		}
+	}
+	if profID == "" {
+		for _, p := range all {
+			if p.NetworkProfile.DomainName == args[0] {
+				profID = p.ID
+				break
+			}
+		}
+	}
+	if profID == "" {
+		for _, p := range all {
+			if p.Name == args[0] {
+				profID = p.ID
+				break
+			}
+		}
+	}
```
If you take the search for a match and put it in a separate func, then this same approach will be much cleaner: you can return whenever a match is found rather than checking that there's still no match after each loop completes.

After you make that change, I think this should be ready.

sailorfrag commented on cmd/tailscale/cli/switch.go:
```diff
@@ -34,6 +34,22 @@ This command is currently in alpha and may change in the future.`,
 	}(),
 	Exec: switchProfile,
+
+	// Add remove subcommand
+	Subcommands: []*ffcli.Command{
```
Hmm, I'd be tempted to make it a flag for the switch command (`tailscale switch -remove <id>`) since that feels more consistent with `-list`, but it sounds like previous conversations suggested it should be a subcommand so I guess that works well enough too. A subcommand is probably more consistent with how other parts of the CLI operate.

@bradfitz is on vacation so I'll just say the subcommand in its current form is fine.

Esteban-Bermudez commented on cmd/tailscale/cli/switch.go:
```diff
@@ -178,3 +194,56 @@ func switchProfile(ctx context.Context, args []string) error {
+		}
+	}
+	if profID == "" {
+		return profID, fmt.Errorf("No profile named %q", arg)
```
I agree that not returning an error makes sense here as we need to distinguish the 2 error situations that could show up. However, do we need to use the comma ok idiom here as `profID` will be `""` every time `ok` is `false`? Is the reason we do this to show the function caller that there is a possibility for `profID` to be empty as without it the function caller might not know to check the value afterwards? 


I think I just answered my own question but would be curious to know for sure.

```go
func matchProfile(arg string, all []ipn.LoginProfile) (ipn.ProfileID, bool) {
	for _, p := range all {
		if p.ID == ipn.ProfileID(arg) {
			return p.ID, true
		}
	}
	for _, p := range all {
		if p.NetworkProfile.DomainName == arg {
			return p.ID, true
		}
	}
	for _, p := range all {
		if p.Name == arg {
			return p.ID, true
		}
	}
	return "", false
}
```

Esteban-Bermudez commented on cmd/tailscale/cli/switch.go:
```diff
@@ -178,3 +194,56 @@ func switchProfile(ctx context.Context, args []string) error {
+	return localClient.DeleteProfile(ctx, profID)
+}
+
+func matchProfile(arg string, cp ipn.LoginProfile, all []ipn.LoginProfile) (ipn.ProfileID, error) {
```
I will add this in as a separate commit to separate the two commands

Esteban-Bermudez commented:
@bradfitz I have rebased with the latest and fixed the merge conflicts. Would it be best to have these changes under the same commit, or keep it separate to indicate that some changes were to fix the switch command while others were for the remove command?
